tldw_Server_API/app/core/File_Artifacts/adapters/ical_adapter.py

```
@@ -0,0 +1,294 @@
+from tldw_Server_API.app.core.File_Artifacts.adapters.base import ExportResult, ValidationIssue
+
+
+class IcalAdapter:
+    file_type: ClassVar[str] = "ical"
+    export_formats: ClassVar[set[str]] = {"ics"}
+    _UTC_EQUIVALENTS: ClassVar[set[str]] = {"UTC", "ETC/UTC", "ETC/GMT", "GMT", "Z"}
+
+    def normalize(self, payload: Dict[str, Any]) -> Dict[str, Any]:
+        calendar = payload.get("calendar")
+        if not isinstance(calendar, dict):
+            raise ValueError("calendar_required")
```

Ruff (0.14.13)

Prefer `TypeError` exception for invalid type

(TRY004)